pkg/keys/bls.go

```diff
@@ -11,44 +11,65 @@ import (
 	"fmt"
 	"io"
 	"io/ioutil"
+	"math/big"
 	"os"
 	"path"
 	"strings"
 
 	ffiBls "github.com/harmony-one/bls/ffi/go/bls"
 	"github.com/harmony-one/go-sdk/pkg/common"
+	"github.com/harmony-one/go-sdk/pkg/sharding"
+	"github.com/harmony-one/go-sdk/pkg/validation"
 	"github.com/harmony-one/harmony/crypto/bls"
 	"github.com/harmony-one/harmony/crypto/hash"
 	"github.com/harmony-one/harmony/shard"
 	"github.com/harmony-one/harmony/staking/types"
 	"golang.org/x/crypto/ssh/terminal"
 )
 
-func GenBlsKeys(passphrase, filePath string) error {
-	privateKey := bls.RandPrivateKey()
-	publicKey := privateKey.GetPublicKey()
-	publicKeyHex := publicKey.SerializeToHexStr()
-	privateKeyHex := privateKey.SerializeToHexStr()
+//BlsKey - struct to wrap bls key data
+type BlsKey struct {
+	PrivateKey     *ffiBls.SecretKey
+	PublicKey      *ffiBls.PublicKey
+	PublicKeyHex   string
+	PrivateKeyHex  string
+	Passphrase     string
+	ShardPublicKey *shard.BlsPublicKey
+}
 
-	if filePath == "" {
-		cwd, _ := os.Getwd()
-		filePath = fmt.Sprintf("%s/%s.key", cwd, publicKeyHex)
-	}
-	if !path.IsAbs(filePath) {
-		return common.ErrNotAbsPath
+// NewBlsKey - create a new bls key struct based on a given private key and passphrase
+func NewBlsKey(privateKey *ffiBls.SecretKey, passphrase string) *BlsKey {
+	publicKey := privateKey.GetPublicKey()
+	blsKey := BlsKey{
+		PrivateKey:    privateKey,
+		PrivateKeyHex: privateKey.SerializeToHexStr(),
+		PublicKey:     publicKey,
+		PublicKeyHex:  publicKey.SerializeToHexStr(),
+		Passphrase:    passphrase,
 	}
-	encryptedPrivateKeyStr, err := encrypt([]byte(privateKeyHex), passphrase)
+
+	return &blsKey
+}
+
+// GenBlsKeys - generate a random bls key using the supplied passphrase, write it to disk at the given filePath
+func GenBlsKeys(passphrase, filePath string) error {
+	blsKey := NewBlsKey(bls.RandPrivateKey(), passphrase)
+	return writeBlsKeyToFile(filePath, blsKey)
+}
+
+// GenMultiBlsKeys - generate multiple BLS keys for a given shard and node/network
+func GenMultiBlsKeys(passphrases []string, filePath string, node string, count uint32, shardID uint32) error {
```

Comment:
any validation conditions to make sense for count? :man_shrugging:

Reply:
Should there be? If they generate too many bls keys for a given network they can't start the node anyways, so the node will validate/set a hard limit anyways. Or are you referring to something else?

Reply:
that was what I was referring to, the limit imposed on running the node. Yeah, if it is imposed when running the node, it seems duplicative.